### DISCUSSION TITLE
Ichimoku Cloud (aka Ichimoku Kinkō Hyō)

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#251] [Indicators]
Community discussion about [Ichimoku Cloud (aka Ichimoku Kinkō Hyō)](https://daveskender.github.io/Stock.Indicators/indicators/Ichimoku/#content)

For calculation, see [Ichimoku Cloud (Wikipedia)](https://en.wikipedia.org/wiki/Ichimoku_Kink%C5%8D_Hy%C5%8D) or wayback on [KumoTrader/IchiWiki](https://web.archive.org/web/20180628174949/http://www.kumotrader.com/ichimoku_wiki/index.php?title=Ichimoku_components).

<img src="https://github.com/DaveSkender/Stock.Indicators/assets/8432125/9067d1e4-a8c7-41cf-b48f-9c3b452a90b3">

**Comment by ijpc1996** [comment]
Hi Dave, very impressive the Ichimoku Cloud. What is the best way to create an image of the result data? Do you recommend a specific library?

**Reply by DaveSkender** [reply]
[Ichimoku Cloud](https://daveskender.github.io/Stock.Indicators/indicators/Ichimoku/#content) is primarily read off of a price overlay chart.  Personally, I've not explored the universe of options of charting libraries.

Let's continue this conversation about charting and data visualization in #430

**Comment by wclark17** [comment]
Why does this function only take 3 parameters while everywhere else takes 4 (of different names, but that is minor).

**Reply by DaveSkender** [reply]
😜 Surely "everywhere" is a bit of an exaggeration, right?  We used the 3 parameter approach that is [also used by StockCharts.com](https://school.stockcharts.com/doku.php?id=technical_indicators:ichimoku_cloud), Interactive Brokers and others.  I see 4 parameters is used by TradingView (and those using TradingView, like BarChart.com, Binance, and a lot of Crypto providers).  If you're familiar with the Ichimoku indicator, it does generally only need 3 parameters; however, it is possible to further customize it.

**Reply by DaveSkender** [reply]
With that said, this is an actively maintained library, so I'm more than happy to add an alternate variant for those who want more dimensions of customization.  Can you provide descriptive examples from the universe of "everywhere" that I can use as a reference?  I've already started thinking of this in #558 

After looking at some popular trading platforms out there, I'm seeing all sorts of different parameters with no consistent naming and almost no descriptions of these extra parameters:
- StockCharts.com uses 3
- Interactive Brokers uses 3 
- Fidelity uses 0 (no parameters)
- MetaTrader uses 3
- ThinkOrSwim uses 2 (or maybe 4?  docs show 2)
- TradingView uses 4
- Tradestation uses 5

**Reply by DaveSkender** [reply]
Ref: [Ichimoku Cloud (Wikipedia)](https://en.wikipedia.org/wiki/Ichimoku_Kink%C5%8D_Hy%C5%8D)

Breaking down Ichimoku a bit more:

**Tenkan-sen** is a signal line, it uses the `signalPeriods` parameter (default=9)
**Kijun-sen** is a high/low confirmation line, it uses the `shortSpanPeriods` parameter (default=26)
**Senkou Span A** is the average of the last two, _offset_ forward by `shortSpanPeriods`
**Senkou Span B** is a high/low over the `longSpanPeriods` parameter (default=52), _offset_ forward by `shortSpanPeriods`
**Chikou Span** is just Close price _offset_ backward by `shortSpanPeriods`

The use of these 3 parameters goes on a common convention that this _offset_ is the same amount as is used for **Kijun-sen** lookback periods.  At least that is my interpretation and a very common interpretation of the original author's intent.

I _think_, adding a fourth parameter breaks this convention by allowing the _offset_ to be specified differently.  Adding a fifth parameter would mean you can have different custom _offset_ for `Senkou` spans and `Chikou` spans.

I think I have some ideas on how to implement these extra options.  Follow design discussions here: #558

**Reply by wclark17** [reply]
Yes, sorry, 'everywhere' Google has been directing. :)  I've been reading https://www.altcointrading.net/ichimoku-cloud/ a bit, and it's the 'displacement', or 4th parameter that I've been working on.  I guess my confusion because I was on FTX which uses Trading View (and on TradingView itself).   Thanks for pointing out the variations, I will check them out as well.

I researched a bit more on the Chikou Span.  In what I read, for the most part the Chikou Span _offset_ is equal to the shortSpanPeriods, but some people trading crypto, as referenced in that link, breaks that equality and uses a separate 'displacement' parameter.   After working on it a bit, I don't know if I agree with those settings.  

I think it would be good to have the parameters, I will follow the discussion!.
Thanks for the quick reply and thanks for a great library!!

**Reply by DaveSkender** [reply]
I've added a 4 and 5 parameter variant in version 1.17.1.  Check it out.  Feedback appreciated.

**Reply by wclark17** [reply]
Will do.  Thanks for this.

Although, now that I played with the settings and seeing exactly what the ‘offset/displacement’ does, I think I like the default.

But appreciate the response and quick turnaround!

**Reply by DaveSkender** [reply]
Yah, I see a lot of debate out there on this indicator around the best use of parameters and values.  I figure I'll just give you the options and let you decide.

**Comment by Salardx** [comment]
Hi,
sorry for asking this, maybe i'm not much familiar with Ichimoku
but when i check this andicator in tradingview i see that this indicator shows SenkouSpan A&B in for the next 26 candles in future
i expected something similar here
but the result of calling GetIchimoku is the same amont of quotes that I provided, 
i can't see any result for future candles

**Reply by DaveSkender** [reply]
![image](https://user-images.githubusercontent.com/8432125/147990370-fb328cf5-69f3-46f7-ac55-f97a1bcacb5e.png)

Yes, it's true that Ichimoku does traditionally expand into the future and this library currently does not.  I got stuck when originally implementing when trying to interpret "the future" from a data perspective.  Does that include holidays and weekends?  And if so, who's holidays and what timezone?  What about hourly and minute data?  Does that start on the next trading day market open?

I think there's potential room for improvement here but might need some help solving this riddle.

**Reply by Salardx** [reply]
I see, if we could find a solution the way you described, it would be more perceise and practical, but at the moment I think we can have a flat view and don't consider any weekend/holiday/timezoon, only next 26 candles would matter (in the exact timeframe of previous candles), I mean although it would seem simpler, but at least this tool would seem more practical, because as you know there are lots of strategies based on future clouds of Ichimoku.
again many thanks for your awesome tools

**Reply by DaveSkender** [reply]
Adding to backlog for tracking: #675

**Reply by UweR70** [reply]
@DaveSkender 
the typically 26 (cloud) "future" values for 
SenkouSpanA (= Leading span A)
and
SenkouSpanB (= Leading span B)
are not related to anthing else than the past chart data!

Please,
get rid of the thought
that something like 
the next week day, holidays, stock market open hours, your size, amount of stars in the universe, or 42
influence these "future" values.

As said they are calculated based on the real chart values of the past - like eveything else in this indicator.

**_Indication that my statement is true:_**
Go for example to "Binance.com"
select a crypto (Doge/EUR),
select the one hour setup 
(this gives you enaugh time to "play" a little bit with the indicator values before the chart will be changed/refreshed!)
add the Ichimoku indicator.

Make sure that the Volume chart is also shown (at the button).

Put the cursor and therefore also the crosshair over the LAST chart value.

Notice that 
a) the crosshair (dotted line) goes also through the volume
b) at the button the currently selected "date and hour(!!!)" 
is shown. 
Example: "13 May '23 18:00" (I am German so I use 24h setup)

Move the mouse SLOWLY to the right.

By doing it slowly you see that the shown date and time(!) changes.

For example : 
It changes 
from "13 May '23 18:00" 
to "13 May '23 19:00" 
etc. etc.

Count the steps until the crosshair is no longer
in the above shown Ichimoku cloud.

You will count exactly 25(!).
(Please ignore that its 25 and not 26!)

Means the "prediction" is 25 "cloud value pairs in the future".

Or in other words:
There are 25 "future" values for 
SenkouSpanA (= Leading span A)
and
SenkouSpanB (= Leading span B)
delivered/shown.

BUT ONLY in case you use the default set up
as follows:
- Conversion Line Periods: 9
- Base Line Periods: 26
- Leasding Span B: 52
- Lagging Span: 26

To demonstrate that my thoughts are correct 
change only
- Lagging Span from 26 to 10!

You will see immedeately that only 9 (nine) 
"cloud value pairs in the future"
are shown.

So please(!)
try to figure out how the
**future**
SenkouSpanA (= Leading span A)
and
SenkouSpanB (= Leading span B)
values are calculated.

Using this indicator 
makes no sense 
in case the 
future cloud values are not provided!

**Comment by hieu-n** [comment]
There might be something wrong with the default offset for Senkou and Chikou. I needed to set them to 25 (instead of 26) to get the correct offset as shown in TradingView chart with default Ichimoku settings which use 26.

Not a big problem for me. I just want to posted here, to see if anyone experiences the same, or maybe it's just my error.

**Reply by DaveSkender** [reply]
When evaluating, please also consider that [TradingView may not be right](https://github.com/DaveSkender/Stock.Indicators/discussions/801).

**Reply by DaveSkender** [reply]
I do think TradingView is incorrect on this one.  If the definition calls for it to “plot results `x` periods ahead”, also phrased as “time-shifted forwards (into the future) `x` periods”, it should result in `x` forward periods, not `x-1`.  Right?

Shift series 1 period, you get 1 future value.
Shift series 5 periods, you get 5 future values.
Shift series 26 periods, you get 26 future values.

If TradingView or Binance have 25 periods in an intended 26 period shift case, it’s very likely wrong.  Either that or they simply drop the last value.

If your goal is to shift 25 places while specifying 26 `kijunPeriods`, you can override by setting `chikouOffset` to 25.

```csharp
var results = quotes.GetIchimoku(9,26,52,26,25);
```
See [our documentation on Ichimoku](https://dotnet.stockindicators.dev/indicators/Ichimoku/#content)

**Reply by DaveSkender** [reply]
There’s also a `senkouOffset` override.

**Comment by UweR70** [comment]
And that´s what chatGPT says:

The ichimoku indicator delivers also cloud values (aka "Leading Span A and B") for the future. How they are calculated?


ChatGPT:

The Ichimoku Cloud indicator is a popular technical analysis tool that uses several lines and areas to provide traders with information about market momentum, trend direction, and potential support and resistance levels.

The two lines that create the Ichimoku Cloud are called the "Leading Span A" and the "Leading Span B". These lines are typically plotted 26 periods ahead of the current price, and are used to identify potential support and resistance levels for future price movements.

The calculation of the Leading Span A and Leading Span B is based on the average of the highest high and the lowest low over a specified number of periods. Here is the step-by-step calculation for each line:

Leading Span A:
- Calculate the average of the 9-period high and 9-period low.
- Calculate the average of the 26-period high and 26-period low.
- Add the results of step 1 and step 2 and divide by 2.
-  Plot the result 26 periods ahead of the current price.

Leading Span B:

- Calculate the average of the 52-period high and 52-period low.
- Plot the result 26 periods ahead of the current price.

Note that the Leading Span A and Leading Span B lines are plotted in the same area on a chart, and the shaded area between them creates the Ichimoku Cloud. The cloud area can be used to identify potential support and resistance levels for future price movements.

**Comment by UweR70** [comment]
Are you looking for this:

1 of 5: Tenkan-sen (Conversion Line): 
This is calculated by adding the highest high and the lowest low over a period of 9 periods 
and then dividing the result by 2.

2 of 5: Kijun-sen (Base Line): 
This is calculated by adding the highest high and the lowest low over a period of 26 periods 
and then dividing the result by 2.

3 of 5: Senkou Span A (Leading Span A): 
This is calculated by adding the Conversion Line and the Base Line 
and then dividing the result by 2. 
The resulting values are plotted 26 periods ahead of the current time period.

4 of 5: Senkou Span B (Leading Span B): 
This is calculated by adding the highest high and the lowest low over a period of 52 periods 
and then dividing the result by 2. 
The resulting values are plotted 26 periods ahead of the current time period.

5 of 5: Chikou Span (Lagging Span): 
This is simply the closing price of the current period plotted 
26 periods behind the current time period.

**Comment by DaveSkender** [comment]
As of right now, the only open issue with this indicator is the **lack of future values** beyond the last price (#675).  The calculation and quantity of future values is known and not an open concern for me, I know how to do that.

I only need to decide on the best date values to apply to those future periods.

Please vote on poll #1060 and continue all discussion related to resolving this specific issue there.  I’d like to get some sentiment information from everyone before deciding on an approach.

**Comment by exogony** [comment]
Hi
Thanks for the package.

It looks like senkou span b uses future values to calculate the current span b. If senkou b periods is 50, the last 50 time points don't have a value, likely indicating that future values are used in the calculation. Am I wrong?

**Reply by DaveSkender** [reply]
Yes, you're wrong.  [Senkou span B](https://en.wikipedia.org/wiki/Ichimoku_Kink%C5%8D_Hy%C5%8D#Senkou_span_B) is calculated with current and prior period values and then simply _shifted_ to future dates.

**Reply by exogony** [reply]
Sorry, I got confused with the names, I meant the chikou span. Based on the description it projects the current price back, so the current chikou_span is the closing price of n forward timepoints. Is that right?

**Reply by DaveSkender** [reply]
Yes, you are right.  Chikou Span pulls a copy of close price backwards.  It’s not actually used to calculate any of the other values, so I wouldn’t call it “repainting”, it’s simply not shown for the last periods until time elapses.  You can see this in our [demo chart](https://charts.stockindicators.dev).

**Reply by AdoNunes** [reply]
nice demo! very useful to visualize all the indicators
